### DISCUSSION TITLE
Device pixel ratio

### DISCUSSION THREAD

**Discussion by fijal** [Kode/Kha#1491] [Q&A]
Hi

I'm very confused about the usage of devicePixelRatio in Kha (and in general about the suspicious lack of examples that don't have some silly fixed size). When I ask it to make 1000x1000 canvas, it makes instead 500x500 canvas. But that's not how the CSS property should be read, I *think* devicePixelRatio = 2 means that we need to draw 2 pixels to get the right idea, as opposed to half a pixel. So bitmaps are small because my device has a large resolution *and* double small because only every second pixel is drawn. What's the plan here? also seems to be hardcoded?

**Comment by RobDangerous** [comment]
Hiho. What exactly do you mean by "ask it to make 1000x1000 canvas"? Kha generally tries to use the actual pixels and all scaling for high dpi displays has to be done manually by the application using Display.pixelsPerInch.

**Comment by fijal** [comment]
If I say something like ``System.start({ title: "GMZ", width: 800, height: 600 }, init);``, it actually makes the canvas half that size because devicePixelRatio is 2. This is the code in ``kha.js``:

```
                                var scale = window.devicePixelRatio;
                                var clientWidth = canvas.clientWidth;
                                var clientHeight = canvas.clientHeight;
                                canvas.width = clientWidth;
                                canvas.height = clientHeight;
                                if(scale != 1) {
                                        canvas.style.width = (clientWidth / scale | 0) + "px";
                                        canvas.style.height = (clientHeight / scale | 0) + "px";
                                }
                                kha_SystemImpl.lastCanvasClientWidth = canvas.clientWidth;
                                kha_SystemImpl.lastCanvasClientHeight = canvas.clientHeight;
```

As far as I can tell it should either not touch the size or make it double (not half)

**Comment by RobDangerous** [comment]
Are you talking about html5 proper or about html5-debug (aka running things in electron)? Proper html5 shouldn't resize at all and ignore the width/height parameters. In html5-debug I remember the situation becoming weirdly complicated.
